storage/storage_image.go

```diff
@@ -90,309 +95,654 @@ func newImageSource(imageRef storageReference) (*storageImageSource, error) {
+	}
+	for _, dataName := range dataNames {
+		bigSize, err := s.transport.store.ImageBigDataSize(s.ID, dataName)
+		if err != nil {
+			return -1, errors.Wrapf(err, "error reading data blob size %q for %q", dataName, s.ID)
+		}
+		sum += bigSize
+	}
+	// Add the signature sizes.
+	for _, sigSize := range s.SignatureSizes {
+		sum += int64(sigSize)
+	}
+	// Prepare to walk the layer list.
+	img, err := s.transport.store.Image(s.ID)
+	if err != nil {
+		return -1, errors.Wrapf(err, "error reading image info %q", s.ID)
+	}
+	// Walk the layer list.
+	layerID := img.TopLayer
+	for layerID != "" {
+		layer, err := s.transport.store.Layer(layerID)
+		if err != nil {
+			return -1, err
+		}
+		if layer.UncompressedSize < 0 {
+			return -1, errors.Errorf("size for layer %q is unknown, failing getSize()", layerID)
+		}
+		sum += layer.UncompressedSize
+		if layer.Parent == "" {
+			break
+		}
+		layerID = layer.Parent
+	}
+	return sum, nil
+}
+
+// newImage creates an image that knows its size and always refers to its layer blobs using
+// uncompressed digests and sizes
+func newImage(s storageReference) (*storageImage, error) {
+	reader, err := newUnnamedImageReader(s)
+	if err != nil {
+		return nil, err
+	}
+	// Compute the image's size.
+	size, err := reader.getSize()
+	if err != nil {
+		return nil, err
+	}
+	// Build the updated information that we want for the manifest.
+	simg, err := reader.transport.store.Image(reader.ID)
+	if err != nil {
+		return nil, err
+	}
+	updatedBlobInfos := []types.BlobInfo{}
+	diffIDs := []digest.Digest{}
+	layerID := simg.TopLayer
+	for layerID != "" {
+		layer, err := reader.transport.store.Layer(layerID)
+		if err != nil {
+			return nil, err
+		}
+		if layer.UncompressedDigest == "" {
+			return nil, errors.Errorf("uncompressed digest for layer %q is unknown", layerID)
+		}
+		blobInfo := types.BlobInfo{
+			Digest: layer.UncompressedDigest,
+			Size:   layer.UncompressedSize,
+		}
+		updatedBlobInfos = append([]types.BlobInfo{blobInfo}, updatedBlobInfos...)
+		diffIDs = append([]digest.Digest{layer.UncompressedDigest}, diffIDs...)
+		if layer.Parent == "" {
+			break
+		}
+		layerID = layer.Parent
+	}
+	info := types.ManifestUpdateInformation{
+		Destination:  nil,
+		LayerInfos:   updatedBlobInfos,
+		LayerDiffIDs: diffIDs,
+	}
+	options := types.ManifestUpdateOptions{
+		LayerInfos:      updatedBlobInfos,
+		InformationOnly: info,
+	}
+	// Return a version of the image that uses the updated manifest.
+	img, err := image.FromSource(reader)
+	if err != nil {
+		return nil, err
+	}
+	updated, err := img.UpdatedImage(options)
+	if err != nil {
+		return nil, err
+	}
+	// Discard a digest in the image reference, if there is one, since the updated image manifest
+	// won't match it.
+	newName := s.name
+	if s.digest != "" {
+		newName = reference.TrimNamed(newName)
+	}
+	newRef := newReference(reader.transport, verboseName(newName), reader.ID, newName, "", "")
```

Comment:
We have to drop the digest, if there is one, to avoid making `UnparsedImage.Manifest` return an error.  Dropping a tag, if that's what we have instead, is kind of a toss-up.  I'd rather not drop either, really.